pkg/mirror/transport.go

```diff
@@ -0,0 +1,257 @@
+package mirror
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/TierMobility/boring-registry/pkg/auth"
+	"github.com/TierMobility/boring-registry/pkg/core"
+
+	"github.com/go-kit/kit/endpoint"
+	httptransport "github.com/go-kit/kit/transport/http"
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+)
+
+type muxVar string
+
+const (
+	varHostname     muxVar = "hostname"
+	varNamespace    muxVar = "namespace"
+	varName         muxVar = "name"
+	varVersion      muxVar = "version"
+	varOS           muxVar = "os"
+	varArchitecture muxVar = "architecture"
+)
+
+type header string
+
+// MakeHandler returns a fully initialized http.Handler.
+func MakeHandler(svc Service, auth endpoint.Middleware, options ...httptransport.ServerOption) http.Handler {
+	r := mux.NewRouter().StrictSlash(true)
+
+	r.Methods("GET").Path(`/{hostname}/{namespace}/{name}/index.json`).Handler(
+		httptransport.NewServer(
+			auth(listProviderVersionsEndpoint(svc)),
+			decodeListVersionsRequest,
+			EncodeJSONResponse,
+			append(
+				options,
+				httptransport.ServerBefore(extractMuxVars(varHostname, varNamespace, varName)),
+				httptransport.ServerBefore(extractHeaders("Authorization")),
+			)...,
+		),
+	)
+
+	r.Methods("GET").Path(`/{hostname}/{namespace}/{name}/{version}.json`).Handler(
+		httptransport.NewServer(
+			auth(listProviderInstallationEndpoint(svc)),
+			decodeListInstallationRequest,
+			EncodeJSONResponse,
+			append(
+				options,
+				httptransport.ServerBefore(extractMuxVars(varHostname, varNamespace, varName, varVersion)),
+				httptransport.ServerBefore(extractHeaders("Authorization")),
+			)...,
+		),
+	)
+
+	r.Methods("GET").Path(`/{hostname}/{namespace}/{name}/terraform-provider-{nameplaceholder}_{version}_{os}_{architecture}.zip`).Handler(
+		httptransport.NewServer(
+			auth(retrieveProviderArchiveEndpoint(svc)),
+			decodeRetrieveProviderArchiveRequest,
+			encodeMirroredResponse,
+			append(
+				options,
+				httptransport.ServerBefore(extractMuxVars(varHostname, varNamespace, varName, varVersion, varOS, varArchitecture)),
+				httptransport.ServerBefore(extractHeaders("Authorization")),
```

Comment:
This can be dropped then.

Reply:
Thanks, addressed this in 4fc34a2e2437abc901387164f0b637c3891a3689